File: portals/ai-workspace/src/paths.ts
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*
 * URL path prefixes this app is wired with.
 *
 * None of these are configuration: each is a fixed contract with the BFF that serves
 * this SPA, so they live here as constants rather than in config.env.ts among the
 * runtime-configurable values. They mirror the BFF's own constants one-for-one
 * (bff/internal/paths/paths.go) and must be changed together with them.
 */

/**
 * URL path prefix this app is served under, with no trailing slash.
 *
 * Single source of truth for the frontend: `vite.config.ts` imports it for Vite's
 * `base`, so the prefix baked into the bundle and the prefix the code composes URLs
 * from cannot disagree. It has to be baked in at all because index.html references its
 * assets by absolute path — a bundle built for one prefix and served under another
 * 404s on every asset. Matches the BFF's `paths.Base`, which mounts every server route.
 *
 * Router paths do NOT need it — BrowserRouter is mounted with basename={BASE_PATH}, so
 * navigate()/<Link to> are already relative to it. It IS needed for anything the router
 * doesn't own: absolute fetch() paths to the BFF, window.location assignments, and OIDC
 * redirect URIs.
 */
export const BASE_PATH = '/ai-workspace';

/**
 * Same-origin prefix every Platform API call goes through: the BFF reverse proxy
 * (<base>/proxy/* → Platform API), so the browser only ever talks to the app origin,
 * never holds a token, and never sees the platform-api self-signed certificate. The
 * BFF strips exactly this prefix before forwarding, so it matches its `paths.Proxy`.
 */
const PROXY_BASE_URL = `${BASE_PATH}/proxy`;

/**
 * Platform API base URL — the proxy base plus the API's own versioned prefix. The
 * version is what this SPA is written against, so reaching a different one is a code
 * change, not a config value. Matches the BFF's `paths.PlatformAPI`.
 */
export const PLATFORM_API_BASE_URL = `${PROXY_BASE_URL}/api/v0.9`;

/**
 * Portal API base URL — the Platform API's portal routes, through the same proxy.
 * Matches the BFF's `paths.PortalAPI`.
 */
export const PORTAL_API_BASE_URL = `${PROXY_BASE_URL}/api/portal/v0.9`;

/**
 * Base URL for the BFF's own API — the routes it answers itself instead of forwarding:
 * session/login/logout, and the handful of creates that span two Platform API calls
 * (secret + resource) and need server-side compensation when the second fails. Callers
 * use it exactly like PLATFORM_API_BASE_URL; which of the two a resource lives under is
 * the only thing that says who handles the request.
 */
export const BFF_API_BASE_URL = `${BASE_PATH}/api`;

File: portals/ai-workspace/bff/internal/paths/paths.go
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// Package paths holds the URL path prefixes this app is wired with. Every one of them
// is a fixed contract — between the BFF and the SPA it ships, or between the BFF and
// the Platform API version it is built against — rather than a deployment concern, so
// they are constants here instead of keys in config.toml. They live in their own
// package because both the server (which routes on them) and config (which composes
// the SPA's runtime config from them) need them.
package paths

// Base is the URL path prefix the whole app is served under: the SPA and its assets,
// the auth endpoints, the runtime-config script and the same-origin proxy all sit
// below it, so https://host:9643/ai-workspace/ is the UI and /ai-workspace/proxy/...
// the API hop. One ingress rule can therefore route a single prefix here with no path
// rewriting, and an all-in-one deployment can put several portals behind one host and
// port without their routes colliding.
//
// The same prefix is baked into the SPA bundle at build time (Vite's `base` in
// vite.config.ts) because index.html references its assets by absolute path — a server
// that moved without a matching rebuild would serve a page whose assets all 404. The
// two must be changed together, and the shipped image is built for this value.
//
// Health is the one route outside the prefix: /healthz answers at the origin root as
// well, so container and Kubernetes probes — which dial the pod directly, bypassing
// the ingress that adds the prefix — keep working.
const Base = "/ai-workspace"

// Proxy is the same-origin reverse-proxy prefix the SPA calls, sitting under Base: the
// browser calls /ai-workspace/proxy/... and the BFF strips both parts before
// forwarding, so the Platform API sees neither and the browser never talks to it
// directly. The SPA receives it pre-composed in its runtime config.
const Proxy = "/proxy"

// PlatformAPI and PortalAPI are the Platform API's versioned route prefixes: the main
// management API, and the portal API under which file-based login (/auth/login) lives.
// They are properties of the upstream this BFF is built against — a different prefix
// means a different API version, which needs code changes here, not a config edit.
const (
	PlatformAPI = "/api/v0.9"
	PortalAPI   = "/api/portal/v0.9"
)
